modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -17,94 +17,156 @@
+import org.apache.ignite.internal.table.distributed.raft.snapshot.message.SnapshotTxDataResponse;
+import org.apache.ignite.internal.tx.storage.state.TxStateStorage;
 import org.apache.ignite.network.ClusterNode;
-import org.apache.ignite.network.MessagingService;
-import org.apache.ignite.network.NetworkMessage;
 import org.apache.ignite.raft.jraft.entity.RaftOutter.SnapshotMeta;
-import org.apache.ignite.raft.jraft.storage.SnapshotStorage;
+import org.apache.ignite.raft.jraft.error.RaftError;
 import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotCopier;
 import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Snapshot copier implementation for partitions. Used to stream partition data from the leader to the local node.
  */
 public class IncomingSnapshotCopier extends SnapshotCopier {
-    /** Messages factory. */
+    private static final IgniteLogger LOG = Loggers.forClass(IncomingSnapshotCopier.class);
+
     private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
 
-    /** {@link SnapshotStorage} instance for the partition. */
-    private final PartitionSnapshotStorage snapshotStorage;
+    private static final long NETWORK_TIMEOUT = 10_000;
 
-    /** Snapshot URI. */
-    private final SnapshotUri snapshotUri;
+    private final PartitionSnapshotStorage partitionSnapshotStorage;
 
-    /** Rebalance thread-pool, used to write data into a storage. */
-    //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-    // Use external pool.
-    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+    private final SnapshotUri snapshotUri;
 
     /**
      * Snapshot meta read from the leader.
      *
      * @see SnapshotMetaRequest
      */
-    private SnapshotMeta snapshotMeta;
+    @Nullable
+    private volatile SnapshotMeta snapshotMeta;
+
+    private volatile boolean canceled;
+
+    @Nullable
+    private volatile CompletableFuture<?> future;
 
     /**
      * Constructor.
      *
-     * @param snapshotStorage Snapshot storage.
+     * @param partitionSnapshotStorage Snapshot storage.
      * @param snapshotUri Snapshot URI.
      */
-    public IncomingSnapshotCopier(PartitionSnapshotStorage snapshotStorage, SnapshotUri snapshotUri) {
-        this.snapshotStorage = snapshotStorage;
+    public IncomingSnapshotCopier(PartitionSnapshotStorage partitionSnapshotStorage, SnapshotUri snapshotUri) {
+        this.partitionSnapshotStorage = partitionSnapshotStorage;
         this.snapshotUri = snapshotUri;
     }
 
     @Override
-    public void cancel() {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement.
+    public void start() {
+        Executor executor = partitionSnapshotStorage.getIncomingSnapshotsExecutor();
+
+        LOG.info("Copier is started for the partition: " + partId());
```

Comment:
Corrected the format, do we need `if (infoEnabled)` if there is a check inside?

Reply:
I vividly remember this topic being discussed a long time ago. The point is that the call without check creates a vararg array argument. It doesn't affect anything in most cases, but sometimes it may become a problem. I didn't invent this, this is how we used to write logs in Ignite 2

Reply:
I agree with you, but we don't have rules for 3.0 yet, maybe we should discuss it and add it to the code style rules.

Reply:
Yes, we should

Reply:
Then, after discussion, a new approach should be applied everywhere.